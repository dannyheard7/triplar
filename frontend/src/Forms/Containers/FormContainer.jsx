import React from "react";
import {inputParsers} from "../utils/inputParsers.js";

export default class FormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            errors: []
        };

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        const data = this.parseFormData(event);

        this.props.apiFunction(data).then(({data}) => {
            let result = data.data.result;

            if (result.errors && result.errors.length > 0) {
                let errors = result.errors.reduce((dict, {field, messages}) => {
                    dict[field] = messages;
                    return dict;
                }, {});

                this.setState({errors: errors})
            } else {
                this.setState({errors: []});
                this.props.onSuccess(result);
            }
        }).catch(error => {
             console.log(error);
        });
    }

    parseFormData(event) {
        const form = event.target;
        const data = new FormData(event.target);

        for (let name of data.keys()) {
            const input = form.elements[name];
            const parserName = input.dataset.parse;

            if (parserName) {
                const parser = inputParsers[parserName];
                const parsedValue = parser(data.get(name));
                data.set(name, parsedValue);
            }
        }

        let variables = Object();
        for(let pair of data.entries()) {
           variables[pair[0]] = pair[1];
        }

        return variables;
    }

    render() {
        const props = {
            onSubmit: this.handleSubmit,
            errors: this.state.errors
        };

        return React.cloneElement(this.props.children, {...props});
    }

}