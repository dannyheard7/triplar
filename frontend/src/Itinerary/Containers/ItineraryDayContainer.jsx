import React from "react";
import ItineraryDay from "Itinerary/Components/ItineraryDay";
import DroppablePlaceListContainer from "Places/Containers/DroppablePlaceListContainer";

import api from "Itinerary/utils/api";

export default class ItineraryDayContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            places: []
        };

        this.getItinerary = this.getItinerary.bind(this);
        this.canDrop = this.canDrop.bind(this);
        this.onDrop = this.onDrop.bind(this);
        this.onDragRemove = this.onDragRemove.bind(this);
    }

    componentDidMount() {
        this.getItinerary();
    }

    getItinerary() {
        api.getLocationDayItinerary(this.props.location.id, this.props.day).then(({data}) => {
            let places = data.data.locationDayItinerary.map(x => x['place']);
            this.setState({places: places});
        });

    }

    canDrop(item) {
        return (this.state.places.indexOf(item) === -1)
    }

    onDrop(item) {
        let places = this.state.places.concat(item);
        this.setState({places: places});

        api.addItemToTripLocation(this.props.location.id, item.id, this.props.day, this.state.places.length);
    }

    onDragRemove(item) {
        let places = this.state.places;
        let index = places.indexOf(item);

        if (index !== -1) {
            places.splice(index, 1);
            this.setState({places: places});
        }
    }

    render () {
        return (
            <div>
                <ItineraryDay day={this.props.day}/>
                <DroppablePlaceListContainer places={this.state.places} path={"/"} canDrop={this.canDrop} onDrop={this.onDrop}
                                             onDragRemove={this.onDragRemove}/>
            </div>
        );
    }
}


