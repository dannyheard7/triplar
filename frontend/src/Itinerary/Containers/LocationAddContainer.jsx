import React from "react";
import DestinationForm from "Itinerary/Components/LocationForm";
import api from "Itinerary/utils/api.js"
import FormContainer from "Forms/Containers/FormContainer";

export default class LocationAddContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showLocationAdd: false
        };

        this.onClick = this.onClick.bind(this);
        this.onSuccess = this.onSuccess.bind(this)
    }

    onClick() {
        this.setState({showLocationAdd: !this.state.showLocationAdd});
    }

    onSuccess(data) {
        this.setState({showLocationAdd: false});
        this.props.onSuccess(data.tripLocation)
    }

    apiFunc = api.curryAPIFunc(api.addLocationToTrip, this.props.tripId);

    render () {
        return (
            <div>
                <button className="btn btn-primary" onClick={this.onClick}>Add Location</button>
                {this.state.showLocationAdd &&
                    <div className="card">
                        <FormContainer onSuccess={this.onSuccess} apiFunction={this.apiFunc}>
                            <DestinationForm submitLabel="Add"/>
                        </FormContainer>
                    </div>
                }
            </div>
        )
    }
}