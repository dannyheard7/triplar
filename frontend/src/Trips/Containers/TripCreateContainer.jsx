import React from "react";
import api from "Trips/utils/trips.api.js"

import TripForm from "Trips/Components/TripForm";
import FormContainer from "../../Forms/Containers/FormContainer";

export default class TripCreateContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onSuccess = this.onSuccess.bind(this);
        this.onClick = this.onClick.bind(this);

        this.state = {
            showTripCreate: false
        };
    }

    onSuccess(data) {
        this.setState({showTripCreate: false});
        this.props.history.push('/trips/' + data.trip.id)
    };

    onClick() {
        this.setState({showTripCreate: !this.state.showTripCreate});
    }

    render () {
        return (
            <div>
                <button className="btn btn-primary" onClick={this.onClick}>New Trip</button>
                {this.state.showTripCreate &&
                    <div className="card">
                        <FormContainer onSuccess={this.onSuccess} apiFunction={api.createTrip}>
                            <TripForm submitLabel="Create" />
                        </FormContainer>
                    </div>
                }
            </div>
        )
    }
}