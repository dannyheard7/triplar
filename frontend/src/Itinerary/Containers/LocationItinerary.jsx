import React from "react";
import Moment from 'moment';
import {connect} from "react-redux";

import {daysBetweenDates} from '../utils/utils.js';
import PlacesSearchContainer from "../../Places/Containers/PlacesSearchContainer";
import ItineraryDayContainer from "./ItineraryDayContainer";
import {addItemToItineraryDay, getItineraryDayDetail, removeItemFromItineraryDay} from "../utils/actions";
import {Helmet} from "react-helmet";
import {DragDropContext} from "react-beautiful-dnd";


export class LocationItinerary extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            selectedDate: props.itinerary.startDate,
            dragging: false,
        };

        this.onChangeSelectedDay = this.onChangeSelectedDay.bind(this);
        this.onDragEnd = this.onDragEnd.bind(this);
        this.onDragStart = this.onDragStart.bind(this);
    }

    componentDidMount() {
        this.props.dispatch(getItineraryDayDetail(this.props.itinerary.id, this.state.selectedDate));
    }

    componentDidUpdate(prevProps) {
        if (this.props.itinerary.id !== prevProps.itinerary.id) {
            this.setState({selectedDate: this.props.itinerary.startDate}, this.componentDidMount);
        }
    }

    onChangeSelectedDay(event) {
        this.setState({selectedDate: event.target.value}, this.componentDidMount);
    }

    onDragEnd(result) {
        this.setState({dragging: false});
        if (!result.destination) return;

        const itineraryId = this.props.itinerary.id;
        let placeId = result.draggableId;
        if(result.source.droppableId === "itinerary-day-droppable") {
            placeId = result.draggableId.slice(0, -itineraryId.toString().length);
        }

        if (result.destination.droppableId === result.source.droppableId) {
            //  reorder
            return;
        }

        const selectedDate = Moment(this.state.selectedDate).format("YYYY-MM-DDz");

        if (result.destination.droppableId === "placeSearchContainer") {
            this.props.dispatch(removeItemFromItineraryDay(itineraryId, placeId, selectedDate));
        } else if (!this.props.itineraryItems.filter(x => x.date === selectedDate).map(x => x.place).includes(placeId)) {
            this.props.dispatch(addItemToItineraryDay(itineraryId, placeId, selectedDate, result.destination.index));
        }
    }

    onDragStart() {
        this.setState({dragging: true});
    }

    render() {
        const {trip, itinerary} = this.props;
        const tripId = trip ? trip.id : this.props.match.params.tripId;
        const path = `/trips/${tripId}/itinerary/${itinerary.id}`;

        const days = daysBetweenDates(itinerary.startDate, itinerary.endDate);

        return (
            <DragDropContext onDragEnd={this.onDragEnd} onDragStart={this.onDragStart}>
                <div>
                    <Helmet>
                        <title>{trip.name + ": " + itinerary.city.name + " | Triplar"}</title>
                    </Helmet>
                    <div>
                        <div className="row">
                            <h3>{itinerary.city.name}</h3>
                            <select onChange={this.onChangeSelectedDay}>
                                {days.map((day) =>
                                    <option key={day} value={day.format("YYYY-MM-DD")}>{day.format('dddd Do MMMM')}</option>
                                )}
                            </select>
                        </div>
                        <ItineraryDayContainer day={this.state.selectedDate} itinerary={itinerary} key={this.state.selectedDate}
                                               path={path} dragging={this.state.dragging}/>
                    </div>
                    <PlacesSearchContainer city={itinerary.city} path={path}/>
                </div>
            </DragDropContext>
        )
    }
}

LocationItinerary.defaultProps = {
    itinerary: {city: {}},
    trip: {}
};

function mapStateToProps(state, ownProps) {
    const itineraryId = ownProps.match.params.itineraryId;
    return {
        itineraryItems: state.itineraries.itineraryItems.filter(x => x.itineraryId === itineraryId),
        itinerary: state.itineraries.locationItineraries.find(x => x.id === itineraryId),
        trip: state.trips.trips.find(x => x.id === ownProps.match.params.tripId)
    };
}

export default connect(mapStateToProps)(LocationItinerary)
