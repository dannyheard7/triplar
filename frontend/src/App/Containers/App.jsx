import React from "react";
import {BrowserRouter, Route, Switch} from "react-router-dom";
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';

import LoggedInContainer from "App/Containers/LoggedInContainer";
import HeaderAuthenticated from "App/Components/HeaderAuthenticated";
import HeaderUnauthenticated from "App/Components/HeaderUnauthenticated";
import Footer from "App/Components/Footer";
import LoginContainer from "Auth/Containers/LoginContainer";
import RegistrationContainer from "Auth/Containers/RegistrationContainer";
import TripListContainer from "Trips/Containers/TripListContainer";
import TripCreateContainer from "Trips/Containers/TripCreateContainer";
import TripDetailContainer from "Trips/Containers/TripDetailContainer";

import 'App/styles/app.css';

export class App extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isAuthenticated: false,
            token: null,
            user: null
        };

        this.onLogin = this.onLogin.bind(this);
        this.onLogout = this.onLogout.bind(this);
    }

    componentDidMount() {
        if (localStorage.getItem("token") !== null && localStorage.getItem("user") !== null) {
          this.setState({isAuthenticated: true, token: localStorage.getItem("token"), user: JSON.parse(localStorage.getItem("user"))})
        }
    }

    onLogin(user, token) {
        this.setState({isAuthenticated: true, token: token, user: user})
    }

    onLogout() {
         this.setState({isAuthenticated: false, token: null, user: null});
    }

    render() {
        if(this.state.isAuthenticated) {
            return (
                <BrowserRouter>
                    <LoggedInContainer token={this.state.token}>
                        <div className="wrapper">
                            <Route path="/" render={(props) => <HeaderAuthenticated {...props} user={this.state.user}
                                                                                    onLogout={this.onLogout} />}/>
                            <div className="content">
                                <Route path="/(|trips)" exact component={TripListContainer}/>
                                <Route path="/(|trips)" exact component={TripCreateContainer}/>
                                <Route path="/trips/:id" component={TripDetailContainer}/>
                            </div>
                            <Footer />
                        </div>
                    </LoggedInContainer>
                </BrowserRouter>
            );
        } else {
            return (
                <BrowserRouter>
                    <div className="wrapper">
                        <HeaderUnauthenticated />
                        <div className="content">
                            <Switch>
                                <Route path="/register" exact component={RegistrationContainer}/>
                                <Route path="/" render={(props) => <LoginContainer {...props} onLogin={this.onLogin} />}/>
                            </Switch>
                        </div>
                        <Footer />
                    </div>
                </BrowserRouter>
            );
        }
    }
}


export default DragDropContext(HTML5Backend)(App)
