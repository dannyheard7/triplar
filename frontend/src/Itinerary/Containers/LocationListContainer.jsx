import React from "react";
import LocationItinerary from "../Components/LocationItinerary";
import LocationAddContainer from "Itinerary/Containers/LocationAddContainer";

import "Itinerary/styles/itinerary.css"

export default class LocationListContainer extends React.Component {
     constructor(props) {
        super(props);

        this.state = {
            locations: this.props.locations
        };

        this.onLocationAdd = this.onLocationAdd.bind(this);
    }

    onLocationAdd(location) {
         this.setState({locations: this.state.locations.concat([location])})
    }

    render() {
        return (
            <div id="trip-itineraries">
                {this.state.locations.map((location) =>
                    <LocationItinerary key={location.id} location={location}/>
                )}
                <LocationAddContainer tripId={this.props.tripId} onSuccess={this.onLocationAdd}/>
            </div>
        );
    }
}