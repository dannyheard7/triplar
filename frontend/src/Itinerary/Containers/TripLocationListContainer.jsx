import React from "react";
import LocationItinerary from "../Components/LocationItinerary";

import "../../Trips/styles/trips.css"
import {connect} from "react-redux";
import {getTripItineraries} from "../../Itinerary/utils/actions";

export class TripLocationListContainer extends React.Component {
    componentDidMount() {
        this.props.dispatch(getTripItineraries(this.props.tripId));
    }

    render() {
        return (
            <div id="trip-itineraries">
                {this.props.locations.map((location) =>
                    <LocationItinerary key={location.id} location={location} tripId={this.props.tripId}/>
                )}
            </div>
        );
    }
}

TripLocationListContainer.defaultProps = {
    locations: []
};

function mapStateToProps(state, ownProps) {
  return {
      locations: state.itineraries.locationItineraries.filter(x => x.trip.id === ownProps.tripId)
  };
}

export default connect(mapStateToProps)(TripLocationListContainer)