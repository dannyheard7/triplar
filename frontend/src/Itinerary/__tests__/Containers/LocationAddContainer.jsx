import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
const renderer = new ShallowRenderer();
const faker = require('faker');

import { TripCreateContainer } from "Trips/Containers/TripCreateContainer";
import FormContainer from "Forms/Containers/FormContainer";
import LocationAddContainer from "Itinerary/Containers/LocationAddContainer";

jest.mock('Itinerary/utils/api.js');

const api = require('Itinerary/utils/api.js');

describe('<LocationAddContainer />', () => {
    const event = {
        preventDefault: () => {},
        target: {
            value: "",
            name: "",
        },
    };

    const data = {tripLocation: {'id': faker.random.number(),
                'city': {'name_std': faker.address.city(), 'country': faker.address.country()}}};


    test('renders correctly', () => {
        const result = renderer.render(<LocationAddContainer tripId="1" />);
        expect(result).toMatchSnapshot();
    });

    test('calls onSuccess after receiving success from <FormContainer />', (function () {
        const spy = jest.spyOn(LocationAddContainer.prototype, 'onSuccess');
        const stub = jest.fn();
        const container = shallow(<LocationAddContainer tripId={faker.random.number()} onSuccess={stub} />);
        container.setState({showLocationAdd: true});
        container.find(FormContainer).prop('onSuccess')(data);

        expect(spy).toBeCalledWith(data);
        expect(stub).toBeCalledWith(data.tripLocation)
    }));

    test('addLocationToTrip called with correct arguments', () => {
        const spy = jest.spyOn(api.default, 'addLocationToTrip');
        const id = faker.random.number();

        const container = shallow(<LocationAddContainer tripId={id} />);
        container.instance().apiFunc(data);

        expect(spy).toBeCalledWith(id, data);

        spy.mockRestore();
    });


    describe('destination create visible', () => {
        let container = null;

        beforeEach(function () {
            container = shallow(<LocationAddContainer tripId={faker.random.number()} onSuccess={jest.fn()}/>);
            container.setState({showLocationAdd: true});
        });

        test('renders <FormContainer /> component when showTripCreate is true', () => {
            expect(container.find(FormContainer).length).toEqual(1);
        });

        test('updates state.showLocationAdd when create button clicked', () => {
            container.find('.btn').simulate('click');
            expect(container.state('showLocationAdd')).toBeFalsy();
        });


        test('does not render <FormContainer /> after a trip has been created', () => {
            expect(container.find(FormContainer).length).toEqual(1);
            container.find(FormContainer).prop('onSuccess')(data);
            container.update();

            expect(container.state('showLocationAdd')).toEqual(false);
            expect(container.find(FormContainer).length).toEqual(0);
        });

    });

});