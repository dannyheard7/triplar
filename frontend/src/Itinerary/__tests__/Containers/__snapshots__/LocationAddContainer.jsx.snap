// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationAddContainer /> renders correctly 1`] = `
<div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Add Location
  </button>
</div>
`;
